src/python/pants/backend/helm/goals/publish_test.py

```diff
@@ -0,0 +1,202 @@
+
+from __future__ import annotations
+
+from typing import cast
+
+import pytest
+
+from pants.backend.helm.goals import publish
+from pants.backend.helm.goals.package import BuiltHelmArtifact
+from pants.backend.helm.goals.publish import HelmPublishFieldSet, PublishHelmChartRequest
+from pants.backend.helm.target_types import HelmChartTarget
+from pants.backend.helm.util_rules import tool
+from pants.backend.helm.util_rules.chart_metadata import HelmChartMetadata
+from pants.backend.helm.util_rules.tool import HelmBinary
+from pants.core.goals.package import BuiltPackage
+from pants.core.goals.publish import PublishPackages, PublishProcesses
+from pants.core.util_rules import external_tool
+from pants.engine.addresses import Address
+from pants.engine.fs import EMPTY_DIGEST
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    return RuleRunner(
+        rules=[
+            *external_tool.rules(),
+            *publish.rules(),
+            *tool.rules(),
+            QueryRule(PublishProcesses, [PublishHelmChartRequest]),
+            QueryRule(HelmBinary, []),
+        ],
+        target_types=[HelmChartTarget],
+    )
+
+
+def _build(metadata: HelmChartMetadata) -> tuple[BuiltPackage, ...]:
+    return (
+        BuiltPackage(
+            EMPTY_DIGEST, (BuiltHelmArtifact.create(f"{metadata.artifact_name}.tgz", metadata),)
+        ),
+    )
+
+
+def _run_publish(
+    rule_runner: RuleRunner,
+    address: Address,
+    metadata: HelmChartMetadata,
+    *,
+    registries: dict | None = None,
+    default_repo: str | None = None,
+) -> tuple[PublishProcesses, HelmBinary]:
+    opts: dict[str, str] = {}
+    opts.setdefault("--helm-registries", "{}")
+
+    if registries:
+        opts["--helm-registries"] = repr(registries)
+    if default_repo:
+        opts["--helm-default-registry-repository"] = default_repo
+
+    rule_runner.set_options([f"{key}={value}" for key, value in opts.items()])
+
+    target = cast(HelmChartTarget, rule_runner.get_target(address))
+    field_set = HelmPublishFieldSet.create(target)
+    packages = _build(metadata)
+
+    result = rule_runner.request(PublishProcesses, [field_set._request(packages)])
+    helm = rule_runner.request(HelmBinary, [])
+    return result, helm
+
+
+def assert_publish(
+    publish: PublishPackages,
+    expect_names: tuple[str, ...],
+    expect_description: str | None,
+    expect_process,
+) -> None:
+    assert publish.names == expect_names
+    assert publish.description == expect_description
+    if expect_process:
+        expect_process(publish.process)
+    else:
+        assert publish.process is None
+
+
+def process_assertion(**assertions):
+    def assert_process(process):
+        for attr, expected in assertions.items():
+            assert getattr(process, attr) == expected
+
+    return assert_process
+
+
+def _declare_targets(rule_runner: RuleRunner) -> None:
```

Comment:
I like this setup, keeping the variations to test together gives a good overview, especially since they're so brief.